### ISSUE TITLE
neg_score 出现Inf导致loss nan

### ISSUE THREAD

**Issue by TangJiakai** [HKUDS/LightGCL#22]
在自定义的数据上出现neg_score 出现Inf导致loss nan的情况，主要bug代码在于下面
https://github.com/HKUDS/LightGCL/blob/5590453ad86782f58017e58d0b698d7f32175be3/model.py#L78-L80

**Comment by L-JuneF** [comment]
这里neg_score 为什么要用```G_u_norm[uids] @ E_u_norm.T```而不是```G_u_norm[uids] @ E_u_norm[uids].T```呢，我记得old版都是两个都是用的批量，这个新版就是用的增强数据的批量和原始图的全部来求neg了

**Comment by TangJiakai** [comment]
> 这里neg_score 为什么要用`G_u_norm[uids] @ E_u_norm.T`而不是`G_u_norm[uids] @ E_u_norm[uids].T`呢，我记得old版都是两个都是用的批量，这个新版就是用的增强数据的批量和原始图的全部来求neg了

那新版会出现这种loss nan的问题，所以说新版这里容易出现数值稳定性的bug，old没试过，那您意思最好用old版本？

**Comment by L-JuneF** [comment]
> > 这里neg_score 为什么要用`G_u_norm[uids] @ E_u_norm.T`而不是`G_u_norm[uids] @ E_u_norm[uids].T`呢，我记得old版都是两个都是用的批量，这个新版就是用的增强数据的批量和原始图的全部来求neg了
> 
> 那新版会出现这种loss nan的问题，所以说新版这里容易出现数值稳定性的bug，old没试过，那您意思最好用old版本？

我不是说最好用old版本，我只是对这两部分的不同有疑问，按照new版本求的neg_score是不是就和论文中的对不上了呢？

**Comment by TangJiakai** [comment]
@L-JuneF 我个人认为batch实现或者全部参与InfoNCE分母计算其实这都可以，这取决于实验效果罢了，但二者都无法回避我提到的问题

**Comment by a-home-bird** [comment]
> @L-JuneF 我个人认为batch实现或者全部参与InfoNCE分母计算其实这都可以，这取决于实验效果罢了，但二者都无法回避我提到的问题

你可以用torchsumexp的API代替，例如，neg_score += (torch.logsumexp(G_i_norm[iids] @ E_i_norm.T / self.temp,dim=1) + 1e-8).mean()

**Comment by L-JuneF** [comment]
> > @L-JuneF我个人认为batch实现或者全部参与InfoNCE分母计算其实这都可以，这取决于实验效果啦，但两者都无法回避我提到的问题
> 
> 你可以用torchsumexp的API代替，例如，neg_score += (torch.logsumexp(G_i_norm[iids] @ E_i_norm.T / self.temp,dim=1) + 1e-8).mean()
有inf问题

**Comment by L-JuneF** [comment]
> @L-JuneF 我个人认为batch实现或者全部参与InfoNCE分母计算其实这都可以，这取决于实验效果罢了，但二者都无法回避我提到的问题

兄弟，你解决了没有？我调小学习率倒是能跑，就是bpr_loss下降十分缓慢

**Comment by TangJiakai** [comment]
> > @L-JuneF 我个人认为batch实现或者全部参与InfoNCE分母计算其实这都可以，这取决于实验效果罢了，但二者都无法回避我提到的问题
> 
> 兄弟，你解决了没有？我调小学习率倒是能跑，就是bpr_loss下降十分缓慢

哎没办法，另外那个哥们提出的方法我还没尝试。你可以把温度系数调大一些吧，lr我倒是没调，0.001,0.005都可以，时间太久了，我也忘了咋改的，反正跑出结果现在没啥问题，收敛速度也还行吧

**Comment by L-JuneF** [comment]
> > > @L-JuneF 我个人认为batch实现或者全部参与InfoNCE分母计算其实这都可以，这取决于实验效果罢了，但二者都无法回避我提到的问题
> > 
> > 
> > 兄弟，你解决了没有？我调小学习率倒是能跑，就是bpr_loss下降十分缓慢
> 
> 哎没办法，另外那个哥们提出的方法我还没尝试。你可以把温度系数调大一些吧，lr我倒是没调，0.001,0.005都可以，时间太久了，我也忘了咋改的，反正跑出结果现在没啥问题，收敛速度也还行吧

代码还能找到吗？我跑ml1m直接nan了

**Comment by TangJiakai** [comment]
@L-JuneF 
原始代码我应该是没变过的，就是纯调参然后跑出来，不过我是自己塔的其他部分pipeline框架，模型部分应该是和新版代码一样的。温度系数=0.5, svd_q=5,lambda_1 = 1e-5, self.lambda_2 = 1e-7, lr=0.001

**Comment by L-JuneF** [comment]
> @L-JuneF 原始代码我应该是没变过的，就是纯调参然后跑出来，不过我是自己塔的其他部分pipeline框架，模型部分应该是和新版代码一样的。温度系数=0.5, svd_q=5,lambda_1 = 1e-5, self.lambda_2 = 1e-7

ok，谢谢，我去调调参试一下
